File: 400/421 - Maximum XOR of Two Numbers in an Array.java
import java.util.HashSet;

public class Solution {
    public int findMaximumXOR(int[] nums) {
        HashSet<Integer> set = new HashSet<Integer>();
        int mask = 0, ret = 0, i, temp;
        
        for (i = 31; i >= 0; --i) {
            mask |= 1 << i;
            temp = ret | (1 << i);
            
            for (int num : nums)
                set.add(num & mask);
        
            for (int b : set) {
                if (set.contains(b ^ temp)) {
                    ret = temp;
                    break;
                }
            }
            set.clear();
        }
        
        return ret;
    }
}

File: csharp/421.maximum-xor-of-two-numbers-in-an-array.cs
/*
 * @lc app=leetcode id=421 lang=csharp
 *
 * [421] Maximum XOR of Two Numbers in an Array
 */

using System.Collections.Generic;

public class Solution {
    public int FindMaximumXOR(int[] nums) {
        ISet<int> set = new HashSet<int>(nums.Length);
        int ret = 0, mask = 0;
        for (int i = 31; i >= 0; --i) {
            mask |= 1 << i;
            foreach (int num in nums) {
                set.Add(num & mask);
            }
            int find = ret | (1 << i);
            foreach (int num in set) {
                if (set.Contains(num ^ find)) {
                    ret = find;
                    break;
                }
            }
            set.Clear();
        }
        return ret;
    }
}

